test/plugin/test_out_http.rb

```diff
@@ -0,0 +1,351 @@
+
+  class << self
+    # Use class variable to reduce server start/shutdown time
+    def startup
+      @@result = nil
+      @@auth_handler = nil
+      @@http_server_thread = nil
+    end
+
+    def shutdown
+      @@http_server_thread.kill
+      @@http_server_thread.join
+    rescue
+    end
+  end
+
+  def server_port
+    19880
+  end
+
+  def base_endpoint
+    "http://127.0.0.1:#{server_port}"
+  end
+
+  def server_config
+    config = {:BindAddress => '127.0.0.1', :Port => server_port}
+    # Suppress webrick logs
+    config[:Logger] = DEFAULT_LOGGER
+    config[:AccessLog] = []
+    config
+  end
+
+  def http_client(**opts, &block)
+    opts = opts.merge(open_timeout: 1, read_timeout: 1)
+    if block_given?
+      Net::HTTP.start('127.0.0.1', server_port, **opts, &block)
+    else
+      Net::HTTP.start('127.0.0.1', server_port, **opts)
+    end
+  end
+
+  def run_http_server
+    server = ::WEBrick::HTTPServer.new(server_config)
+    server.mount_proc('/test') { |req, res|
+      if @@auth_handler
+        @@auth_handler.call(req, res)
+      end
+
+      @@result.method = req.request_method
+      @@result.content_type = req.content_type
+      req.each do |key, value|
+        @@result.headers[key] = value
+      end
+
+      data = []
+      case req.content_type
+      when 'application/x-ndjson'
+        req.body.each_line { |l|
+          data << JSON.parse(l)
+        }
+      when 'text/plain'
+        # Use single_value in this test
+        data = req.body.lines(chomp: true)
+      else
+        data << req.body
+      end
+      @@result.data = data
+
+      res.status = 200
+      res.body = "success"
+    }
+    server.mount_proc('/503') { |_, res|
+      res.status = 503
+      res.body = 'Service Unavailable'
+    }
+    server.mount_proc('/404') { |_, res|
+      res.status = 404
+      res.body = 'Not Found'
+    }
+    # For start check
+    server.mount_proc('/') { |_, res|
+      res.status = 200
+      res.body = 'Hello Fluentd!'
+    }
+    server.start
+  ensure
+    server.shutdown rescue nil
+  end
+
+  Result = Struct.new("Result", :method, :content_type, :headers, :data)
+
+  setup do
+    Fluent::Test.setup
+    FileUtils.rm_rf(TMP_DIR)
+
+    @@result = Result.new(nil, nil, {}, nil)    
+    @@http_server_thread ||= Thread.new do
+      run_http_server
+    end
+
+    now = Time.now
+    started = false
+    until started
+      raise "Server not started" if (now - Time.now > 10.0)
+      begin
+        http_client { |c| c.request_get('/') }
+        started = true
+      rescue
+        sleep 0.5
+      end
+    end
+  end
+
+  teardown do
+    @@result = nil
+    @@auth_handler = nil
+  end
+
+  def create_driver(conf)
+    Fluent::Test::Driver::Output.new(Fluent::Plugin::HTTPOutput).configure(conf)
+  end
+
+  def config
+    %[
+      endpoint #{base_endpoint}/test
+    ]
+  end
+
+  def test_events
+    [
+      {"message" => "hello", "num" => 10, "bool" => true},
```

Comment:
[nits]To use a newer version hash is better.

Reply:
What does "newer version" mean?
`{message: "hello"}` creates symbol key, not string key.

Reply:
`{"message": "hello", "num": 10, "bool": true}` style hash does not handle Ruby 2.1.
This feature had been provided Ruby 2.2 or later.
For now, we shouldn't drop Ruby 2.1 support.
So, hash rocket notation is needed here.

Reply:
Ah, string key is needed here. I see 👌